### PR TITLE
feat: add support for python 3.12

### PR DESCRIPTION
**Pull request by Yun-Kim**
Adds support for Python 3.12.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: tests/debugging/py35/test_async.py
Comment (P403n1x87) [inline tests/debugging/py35/test_async.py:1]: Apologies if I have overlooked something, but it looks like these tests under a `pyXY` folder have been removed but not migrated to the `tests/debugging` folder. The `pyXY` folder means "Python X.Y and above", so we still need to run these tests.

Comment (P403n1x87) [inline tests/debugging/py35/test_async.py:1]: Apologies if I have overlooked something, but it looks like these tests under a `pyXY` folder have been removed but not migrated to the `tests/debugging` folder. The `pyXY` folder means "Python X.Y and above", so we still need to run these tests.

**Review by P403n1x87 — DISMISSED** [review]
I think we need to make sure we don't remove tests that are still relevant

File: tests/debugging/py35/test_async.py
Comment (Yun-Kim) [reply tests/debugging/py35/test_async.py:1]: Ahh I was under the assumption that `pyXY` meant Python X.Y specific only, I'll re-add those tests.

Comment (Yun-Kim) [reply tests/debugging/py35/test_async.py:1]: @majorgreys we might need to revert your commit removing the py35/36 files, context above 😄

Comment (majorgreys) [reply tests/debugging/py35/test_async.py:1]: Discussed with @Yun-Kim offline. 

We keep the debugger tests but just move them to an appropriately named folder:

https://github.com/DataDog/dd-trace-py/pull/6626/commits/fcb684531b76a72900731e90573a7828cb4cc6a5#diff-f4025bf6849aa4878550279f03e198ff93377462eb6679841b47d8745ad35587.

@P403n1x87 can you review?

Comment (P403n1x87) [reply tests/debugging/py35/test_async.py:1]: I think we no longer need to keep the `pyXY` folder structure, so we could move/merge those scripts into the parent folder.

File: .riot/requirements/releasenotes/notes/3.12_support-422a62d1871877c2.yaml
Comment (emmettbutler) [inline .riot/requirements/releasenotes/notes/3.12_support-422a62d1871877c2.yaml:3]: This needs more information

File: .riot/requirements/releasenotes/notes/3.12_support-422a62d1871877c2.yaml
Comment (emmettbutler) [inline .riot/requirements/releasenotes/notes/3.12_support-422a62d1871877c2.yaml:10]: This is the default text and should be removed or changed

File: hatch.toml
Comment (emmettbutler) [inline hatch.toml:32]:
```suggestion
    "riot==0.19.0",
```

File: ddtrace/tracer.py
Comment (emmettbutler) [inline ddtrace/tracer.py:80]:
```suggestion
```

File: .riot/requirements/releasenotes/notes/3.12_support-422a62d1871877c2.yaml
Comment (emmettbutler) [inline .riot/requirements/releasenotes/notes/3.12_support-422a62d1871877c2.yaml:7]: This should probably include a mention of dropped support for <3.8

File: ddtrace/__init__.py
Comment (emmettbutler) [inline ddtrace/__init__.py:21]:
```suggestion
# configure ddtrace logger before other modules log
```

File: ddtrace/__init__.py
Comment (majorgreys) [inline ddtrace/__init__.py:57]: Added above 

```suggestion
```

File: .riot/requirements/releasenotes/notes/3.12_support-422a62d1871877c2.yaml
Comment (Yun-Kim) [reply .riot/requirements/releasenotes/notes/3.12_support-422a62d1871877c2.yaml:7]: We added a separate release note for dropping support Python < 3.7 here: https://github.com/DataDog/dd-trace-py/blob/yunkim/add-py-312/releasenotes/notes/release-2.0-3af0045e2261bd02.yaml

File: ddtrace/debugging/__init__.py
Comment (majorgreys) [inline ddtrace/debugging/__init__.py:29]:
```suggestion
```

**Review by emmettbutler — APPROVED** [review]

**Review by emmettbutler — APPROVED** [review]

**Review by Kyle-Verhoog — COMMENTED** [review]
We need to update the library injection image to also build and include Python 3.12: https://github.com/DataDog/dd-trace-py/blob/bb8f196591fbfc8476c1fbf00787d9dcf52026ac/lib-injection/Dockerfile#L13

**Comment by emmettbutler** [comment]
@Kyle-Verhoog done!

**Comment by emmettbutler** [comment]
@Kyle-Verhoog let's address that in a separate pull request if that's ok with you

**Review by emmettbutler — DISMISSED** [review]

**Comment by emmettbutler** [comment]
@Kyle-Verhoog let's make that change in a separate pull request if that's ok with you.

**Comment by Kyle-Verhoog** [comment]
@emmettbutler that's fine with me!

**Review by emmettbutler — DISMISSED** [review]

**Review by shatzi — DISMISSED** [review]

**Review by ZStriker19 — DISMISSED** [review]

**Comment by emmettbutler** [comment]
commit https://github.com/DataDog/dd-trace-py/pull/6626/commits/bc1eec51b16c27b6e22784c657648e38ba93b4b5 broke the opentelemetry test suite while fixing the other failures

**Review by emmettbutler — DISMISSED** [review]

**Comment by emmettbutler** [comment]
@ZStriker19 @P403n1x87 @majorgreys @shatzi @Kyle-Verhoog This is actually ready for review now, with all tests passing. Please approve again at your convenience!

File: ddtrace/internal/safety.py
Comment (P403n1x87) [inline ddtrace/internal/safety.py:70]: Interesting 🤔. Why doesn't this work with CPython 3.12?
```suggestion
        if PYTHON_VERSION_INFO < (3, 12) and name == "__wrapped__":
```

File: setup.py
Comment (P403n1x87) [inline setup.py:587]:
```suggestion
    setup_requires=["setuptools_scm[toml]>=4", "cython", "cmake>=3.24.2"],
```

**Review by P403n1x87 — COMMENTED** [review]
A couple of nits and questions. I think this needs a new rebase, otherwise LGTM!

File: ddtrace/internal/safety.py
Comment (emmettbutler) [reply ddtrace/internal/safety.py:70]: To be honest I'm not certain. I suspect it has to do with this change in the latest version of wrapt https://github.com/GrahamDumpleton/wrapt/commit/dd81bd7b5b604c58572c4863c25bdd2385d56468

**Comment by emmettbutler** [comment]
@P403n1x87 thanks for checking this out. what do you mean by "needs a new rebase"?

**Review by emmettbutler — APPROVED** [review]

**Comment by emmettbutler** [comment]
@P403n1x87 ah, thanks for clarifying. do you know of any reason that needs to happen before this gets merged?

**Review by emmettbutler — APPROVED** [review]

**Review by P403n1x87 — APPROVED** [review]

**Pull request merged**
